Implement HpkeDecryptor functionality for MockAggregator

nakatsuka-y opened the pull request:
This is the start of the process of addressing issue #20.

As for the first step, these couple of commits implements the HpkeDecryptor functionality for MockAggregator.
The test code for the HpkeDecryptor functionality is also added.

armfazh commented on daphne/src/roles_test.rs:
```diff
@@ -318,3 +366,70 @@ async fn http_post_collect_unauthorized_request() {
 }
+
+#[test]
+fn hpke_decrypter() {
```
In tests that use a mock object, usually their methods are not invoked directly, but by other piece of code. Otherwise, this only tests that the mock was well-implemented, but not that is does it's job when used by other party.

nakatsuka-y replied:
The test code I wrote there is based on the other tests in the same file.
MockAggregator is not used by other code yet, so I would need to add some code that uses MockAggregator.

armfazh reviewed (approved):
lgtm, just let's see how the mocks can be used elsewhere

nakatsuka-y commented:
I've added a new test case that ultimately uses [`cosume_report_share`](https://github.com/cloudflare/daphne/blob/82bbcc1a39258a7a3b11476ebc1a72e5d993200f/daphne/src/vdaf/mod.rs#L186), which must decrypt the input share data to consume it.

The check is based on whether the decryption fails.
I am currently working on a test case to check whether the decryption succeeds.

nakatsuka-y commented:
I've implemented a test case where the decryption succeeds.
Hopefully the two test cases I've added in the past few commits allows others to understand how the mock_backend can be used.

nakatsuka-y commented on daphne/src/roles.rs:
```diff
@@ -417,6 +417,7 @@ pub trait DapHelper<S>: DapAggregator<S> {
 
     /// Handle an HTTP POST to `/aggregate`. The input is an AggregateInitializeReq and the
```
The comment has been updated to add `AggregateContinueReq`.

nakatsuka-y commented on daphne/src/roles.rs:
```diff
@@ -433,6 +434,7 @@ pub trait DapHelper<S>: DapAggregator<S> {
                     .mark_aggregated(&agg_init_req.task_id, &agg_init_req.report_shares)
                     .await?;
 
+                // This is where the encrypted report share is decrypted
```
All comments that do not have proper punctuation have been updated.

nakatsuka-y commented on daphne/src/roles_test.rs:
```diff
@@ -28,27 +29,67 @@ const TASK_LIST: &str = r#"{
-                    "buckets": [77, 999]
-                }
-            }
+            "prio3": "count"
```
This is because we requesting a `Prio3Config::Count` measurement.
(ref: https://github.com/nakatsuka-y/daphne/blob/1cecf456be35cb11a3af82bd67441a60c4e335db/daphne/src/roles_test.rs#L437)

nakatsuka-y commented on daphne/src/roles_test.rs:
```diff
@@ -28,27 +29,67 @@ const TASK_LIST: &str = r#"{
+        "id": 14,
+        "sk": "b809a4df399548f56c3a15ebaa4925dd292637f0b7e2f6bc3ba60376b69aa05e"
+    }
+]"#;
```
I was wondering whether there was any reason why `Copy` or `Clone` was not added (e.g., to prevent any copying / cloning of secret keys?).
In any case, I can implement that and the `HpkeReceiverConfig` struct.

cjpatton replied:
As a general rule it's best to avoid implementing functionality that's not exercised. Typically if you need `Clone`, you should feel free to add it :)

nakatsuka-y replied:
I've implemented the `HpkeReceiverConfig` struct and used it to replace the `HPKE_CONFIG_LIST` and `HPKE_SECRET_KEY_LIST` in `roles_test.rs`.
There are several changes needed to support this, mainly adding `#[serde(with = "hex")]` to the `HpkeConfig` struct so that serde is able to parse public keys encoded in hex.

nakatsuka-y commented on daphne/src/roles.rs:
```diff
@@ -492,8 +495,10 @@ pub trait DapHelper<S>: DapAggregator<S> {
-    /// Handle an HTTP POST to `/aggregate_share`. The input is an AggregateShareReq and the
-    /// response is an AggregateShareResp.
+    /// Handle an HTTP POST to `/aggregate_share`.
+    /// The input is an AggregateShareReq and the response is an AggregateShareResp.
```
The recent pushes should reflect this feedback.

cjpatton commented on daphne/src/roles_test.rs:
```diff
@@ -318,3 +392,169 @@ async fn http_post_collect_unauthorized_request() {
+    let transition = &agg_resp.transitions[0];
+
+    // Expect failure due to invalid ciphertext.
+    assert_matches!(transition.var, TransitionVar::Failed(_));
```
We should assert the type of `TransitionFailure` here as well.

cjpatton reviewed (commented):
Coming along nicely! A few more minor comments in-line.

nakatsuka-y commented on daphne/src/roles_test.rs:
```diff
@@ -28,27 +29,83 @@ const TASK_LIST: &str = r#"{
     }
 }"#;
 
+const HPKE_RECEIVER_CONFIG_LIST: &str = r#"[
```
I was thinking of removing the redundant `id` field, so I completely agree with this comment.
I'll open a new PR that resolves this.

nakatsuka-y commented on daphne/src/hpke.rs:
```diff
@@ -148,3 +148,10 @@ pub trait HpkeDecrypter {
 }
+
+/// Struct that combines HpkeConfig and HpkeSecretKey
+#[derive(Clone, Deserialize, Serialize)]
```
I've removed `Clone` from both `HpkeReceiverConfig` and `HpkeSecretKey`.

nakatsuka-y commented on daphne/src/roles_test.rs:
```diff
@@ -77,21 +134,33 @@ impl BearerTokenProvider for MockAggregator {
+            match hpke_secret_key.decrypt(info, aad, &ciphertext.enc, &ciphertext.payload) {
+                Ok(plaintext) => Ok(plaintext),
+                Err(_) => Err(DapError::Transition(TransitionFailure::HpkeDecryptError)),
+            }
```
This change was made when I didn't know that there was a translation from `HpkeError` to `TransitionFailure`.
I've changed this to use `?`.

nakatsuka-y commented on daphne/src/roles_test.rs:
```diff
@@ -318,3 +392,169 @@ async fn http_post_collect_unauthorized_request() {
+    let task_id = helper.nominal_task_id();
+    let task_config = helper.get_task_config_for(task_id).unwrap();
+
+    // Construct Hpke receiver config List.
```
Changed `Hpke` to `HPKE` throughout the comments.

nakatsuka-y commented on daphne/src/roles_test.rs:
```diff
@@ -149,7 +218,10 @@ impl DapHelper<BearerToken> for MockAggregator {
     ) -> Result<HashMap<Nonce, TransitionFailure>, DapError> {
-        unreachable!("not implemented");
+        // Return empty HashMap (for now).
+        // TODO: Implement correct functionality.
```
I've add my username to all TODO notes.

cjpatton reviewed (commented):
Alright, this is ready to merge. Before we do, please "squash" all of the commits into a single commit and make sure the commit message reflects the change.

```
git rebase main -i
```

This command is used to rebase "interactively". For each commit, you can choose an action for applying it. You'll want to "pick" the commit at the bottom of the stack, but "squash" each subsequent commit into the previous one.
